crates/vm/src/system/memory/controller/memory.rs

```diff
@@ -406,17 +487,142 @@ impl<F: PrimeField32> Memory<F> {
             .map(|i| self.get(address_space, pointer + i as u32))
             .collect()
     }
+
+    pub fn aux_cols_factory(&self) -> MemoryAuxColsFactory<F> {
+        let range_bus = self.range_checker.bus();
+        MemoryAuxColsFactory {
+            range_checker: self.range_checker.clone(),
+            timestamp_lt_air: AssertLtSubAir::new(range_bus, self.timestamp_max_bits),
+            _marker: Default::default(),
+        }
+    }
+
+    // just for unit testing
+    #[cfg(test)]
+    fn last_record(&self) -> &MemoryRecord<F> {
+        self.log.last().unwrap().as_ref().unwrap()
+    }
+}
+
+pub type MemoryImage<F> = FxHashMap<Address, F>;
+
+/// A simple data structure to read to/write from memory.
+///
+/// Stores a log of memory accesses to reconstruct aspects of memory state for trace generation.
+#[derive(Debug, Clone)]
+pub struct Memory<F> {
+    pub(super) data: FxHashMap<Address, F>,
+    pub(super) log: Vec<MemoryLogEntry<F>>,
+    timestamp: u32,
+}
+
+impl<F: PrimeField32> Default for Memory<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<F: PrimeField32> Memory<F> {
+    pub fn new() -> Self {
+        Self {
+            data: MemoryImage::default(),
+            timestamp: INITIAL_TIMESTAMP + 1,
+            log: vec![],
+        }
+    }
+
+    /// Instantiates a new `Memory` data structure from an image.
+    pub fn from_image(image: MemoryImage<F>) -> Self {
+        Self {
+            data: image,
+            timestamp: INITIAL_TIMESTAMP + 1,
+            log: vec![],
+        }
+    }
+
+    fn last_record_id(&self) -> RecordId {
+        RecordId(self.log.len() - 1)
+    }
+
+    /// Writes an array of values to the memory at the specified address space and start index.
+    ///
+    /// Returns the `RecordId` for the memory record and the previous data.
+    pub fn write<const N: usize>(
+        &mut self,
+        address_space: u32,
+        pointer: u32,
+        values: [F; N],
+    ) -> (RecordId, [F; N]) {
+        assert!(N.is_power_of_two());
+
+        let prev_data = array::from_fn(|i| {
+            self.data
+                .insert((address_space, pointer + i as u32), values[i])
+                .unwrap_or(F::ZERO)
+        });
+
+        self.log.push(MemoryLogEntry::Write {
+            address_space,
+            pointer,
+            data: values.to_vec(),
```

Comment:
I'm a bit concerned about this `to_vec()` all the time. maybe we just need to make the enum have a bunch of powers of two variants